### ISSUE TITLE
DistanceBand fails to accept custom ids

### ISSUE THREAD

**Issue by ljwolf** [pysal/pysal#810]
`pysal.DistanceBand(points, threshold, ids=ids)` fails to construct weights when reindexing the resulting weights object:

```python
>>> import pysal as ps
>>> import numpy as np
>>> shp = ps.open(ps.examples.get_path('columbus.shp'))
>>> XY = np.array([list(x.centroid) for x in shp])
>>> dbf = ps.open(ps.examples.get_path('columbus.dbf'))
>>> ids = dbf.by_col('POLYID')
>>> ps.DistanceBand(XY, 5, ids=ids)
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-7-d9532ce435ec> in <module>()
----> 1 ps.DistanceBand(XY, 5, ids=ids)

/home/ljw/dev/pysal/pysal/weights/Distance.pyc in __init__(self, data, threshold, p, alpha, binary, ids)
    472         self.alpha = alpha
    473         self._band()
--> 474         neighbors, weights = self._distance_to_W(ids)
    475         W.__init__(self, neighbors, weights, ids)
    476 

/home/ljw/dev/pysal/pysal/weights/Distance.pyc in _distance_to_W(self, ids)
    496                         weights[i].append(1)
    497                         neighbors[i].append(j)
--> 498                     if i not in neighbors[j]:
    499                         weights[j].append(1)
    500                         neighbors[j].append(i)

KeyError: 0
```
This is independent of python version, but is being replicated with pysal 1.11.1, numpy 1.10.4, scipy 0.17.1.

**Comment by ljwolf** [comment]
Also, it happens regardless of the distance band. I just used 5 arbitrarily :sweat_smile:

**Comment by ljwolf** [comment]
I believe this occurs because, `self.dmat` is a matrix with linear indexes, but `neighbors` and `weights` are constructed with arbitrary indices. So, when we iterate through `self.dmat.items()` (linearly 0-indexed), we canot find those in the arbitrarily-indexed collections we're building.

**Comment by ljwolf** [comment]
For reference, this in the [`_band` and `_distance_to_W`](https://github.com/pysal/pysal/blob/master/pysal/weights/Distance.py#L367) functions. 

![2016-05-25-103449_937x1194_scrot](https://cloud.githubusercontent.com/assets/2250995/15549876/8f035d3e-2263-11e6-9f02-0dde1dd9f30e.png)

**Comment by ljwolf** [comment]
closed in #811
